spacy/ml/models/entity_linker.py

```diff
@@ -29,6 +31,52 @@ def build_nel_encoder(
     return model
 
 
+def build_span_maker(n_sents: int = 0) -> Model:
+    model: Model = Model("candidate_docs", forward=span_maker_forward)
+    model.attrs["n_sents"] = n_sents
+    return model
+
+
+def span_maker_forward(model, docs: List[Doc], is_train) -> Tuple[Ragged, Callable]:
+    ops = model.ops
+    n_sents = model.attrs["n_sents"]
+    candidates = []
+    for doc in docs:
+        cands = []
+        try:
+            sentences = [s for s in doc.sents]
+        except ValueError:
+            # no sentence info, normal in initialization
+            for tok in doc:
+                tok.is_sent_start = tok.i == 0
+            sentences = [doc[:]]
+        for ent in doc.ents:
+            try:
+                # find the sentence in the list of sentences.
+                sent_index = sentences.index(ent.sent)
+            except AttributeError:
+                # Catch the exception when ent.sent is None and provide a user-friendly warning
+                raise RuntimeError(Errors.E030) from None
+            # get n previous sentences, if there are any
+            start_sentence = max(0, sent_index - n_sents)
+            # get n posterior sentences, or as many < n as there are
+            end_sentence = min(len(sentences) - 1, sent_index + n_sents)
+            # get token positions
+            start_token = sentences[start_sentence].start
+            end_token = sentences[end_sentence].end
+            # save positions for extraction
+            cands.append((start_token, end_token))
+
+        candidates.append(ops.asarray2i(cands))
+    candlens = ops.asarray1i([len(cands) for cands in candidates])
+    candidates = ops.xp.concatenate(candidates)
+    outputs = Ragged(candidates, candlens)
+    # TODO W093 if there's no data. Can we do that here though?
+
+    # because this is just rearranging docs, the backprop does nothing
+    return outputs, lambda x: []
```

Comment:
But wait - if we don't backprop back, the `tok2vec` layer won't receive gradients and won't be able to tune itself towards the EL task?

Reply:
I believe that's not relevant here because this is parallel to the tok2vec, not after it. 

```
        model = (
            ((tok2vec >> list2ragged()) & build_span_maker())
            >> extract_spans()
            >> reduce_mean()
            >> residual(Maxout(nO=token_width, nI=token_width, nP=2, dropout=0.0))  # type: ignore[arg-type]
            >> output_layer
        )
```

The `&` is tuplify, so `extract_spans` will backprop to `tok2vec >> list2ragged()` as well as the span maker. But the span maker has no statistical anything - it just generates indices for `extract_spans()` - so it doesn't need to do backprop.

Does that make sense or am I missing something?

Reply:
Oh yea you're probably right! Hm that still leaves us with the issue of the `nan`'s though. When you print out the internal values at the various input/output steps of the different layers, you can often see 0's creeping into the vectors, eventually zero'ing the entire network.